ports/stm32f4/common-hal/busio/UART.c

```diff
@@ -110,8 +108,8 @@ void common_hal_busio_uart_construct(busio_uart_obj_t *self,
         if(self->tx!=NULL && self->rx!=NULL) {
             USARTx = mcu_uart_banks[self->tx->uart_index-1];
             mp_printf(&mp_plat_print, "UART:%d \n", self->tx->uart_index);
-            //TODO: remove this horrible hack 
-            context_pointers[self->tx->uart_index-1] = self;
+            //assign a root pointer pointer for IRQ
```

Comment:
Please factor out this chunk. It's in here three times.

Reply:
Are you sure? it's pretty ugly and it obscures functionality 
```
STATIC USART_TypeDef * assign_uart_or_throw(busio_uart_obj_t *self, bool pin_eval, 
                                            int uart_index, bool uart_taken) {
    if(pin_eval) {
        //assign a root pointer pointer for IRQ
        MP_STATE_PORT(cpy_uart_obj_all)[uart_index] = self;
        return mcu_uart_banks[uart_index];
    } else {
        if (uart_taken) {
            mp_raise_ValueError(translate("Hardware busy, try alternative pins"));
        } else {
            mp_raise_ValueError(translate("Invalid UART pin selection"));
        }
    }
}
```

Reply:
Implemented: check it out and see if you prefer.

Reply:
I agree it's a bit weird but I like it more than having it repeated. The other option is to move it after the three pin searches.